pkg/controller/controller_broker.go

```diff
@@ -391,28 +404,15 @@ func (c *controller) reconcileClusterServiceBroker(broker *v1alpha1.ClusterServi
 	if finalizers := sets.NewString(broker.Finalizers...); finalizers.Has(v1alpha1.FinalizerServiceCatalog) {
 		glog.V(4).Infof("ClusterServiceBroker %q: finalizing", broker.Name)
 
-		fieldSet := fields.Set{
-			"spec.clusterServiceBrokerName": broker.Name,
-		}
-		fieldSelector := fields.SelectorFromSet(fieldSet).String()
-		listOpts := metav1.ListOptions{FieldSelector: fieldSelector}
-
-		svcPlans, err := c.serviceCatalogClient.ClusterServicePlans().List(listOpts)
+		existingServiceClasses, existingServicePlans, err := c.getCurrentServiceClassesAndPlansForBroker(broker)
 		if err != nil {
-			c.updateClusterServiceBrokerCondition(
-				broker,
-				v1alpha1.ServiceBrokerConditionReady,
-				v1alpha1.ConditionUnknown,
-				errorListingClusterServicePlansReason,
-				errorListingClusterServicePlansMessage,
-			)
-			c.recorder.Eventf(broker, corev1.EventTypeWarning, errorListingClusterServicePlansReason, "%v %v", errorListingClusterServicePlansMessage, err)
 			return err
 		}
 
-		glog.V(4).Infof("ClusterServiceBroker %q: found %d ClusterServicePlans to delete", broker.Name, len(svcPlans.Items))
+		glog.V(4).Infof("ClusterServiceBroker %q: found %d ClusterServiceClasses to delete", broker.Name, len(existingServiceClasses))
+		glog.V(4).Infof("ClusterServiceBroker %q: found %d ClusterServicePlans to delete", broker.Name, len(existingServicePlans))
 
-		for _, plan := range svcPlans.Items {
+		for _, plan := range existingServicePlans {
```

Comment:
Here might add a comment saying that everything left in existingServicePlans/existingServiceMaps is no longer on the Broker side so you remove them.

Reply:
yep, totes agree.